pkg/cache/cluster.go

```diff
@@ -1192,3 +1230,89 @@ func (c *clusterCache) GetClusterInfo() ClusterInfo {
 func skipAppRequeuing(key kube.ResourceKey) bool {
 	return ignoredRefreshResources[key.Group+"/"+key.Kind]
 }
+
+// UpdateClusterConnectionStatus starts a goroutine that checks for watch failures.
+// If there are any watch errors, it will periodically ping the remote cluster
+// and update the cluster connection status.
+func (c *clusterCache) UpdateClusterConnectionStatus(ctx context.Context) {
+	go c.clusterConnectionService(ctx)
+}
+
+func (c *clusterCache) clusterConnectionService(ctx context.Context) {
+	clusterConnectionTimeout := 10 * time.Second
```

Comment:
I've used 10 seconds to poll the status of watch failures. We might have to ping the remote cluster every 10s if the watches are failing. But the function `updateConnectionStatus` will run only once when the status changes. For the subsequent polls, the condition checks ensure we don't call it often. 

These are the conditions that we check before invalidating the cache:
    1. Check if there are watch failures or if it has recovered back(no watch errors but the connection status is 'Failed')
    2. If either of the above conditions are met we ping the remote cluster to get the version.
    3. If there is no error, but the connection status is `Failed` we invalidate the cache.
    4. If there is an error, but the connection status is `Successful` we invalidate the cache.
    
I'm open to updating to a longer duration. But the tradeoff is we can't recognize the failures early. This should be okay I suppose, since Argo CD will not be primarily used for cluster monitoring.

Reply:
I'd recommend making the interval configurable per cache, so it can be parametrized from a consumer

Reply:
Made the interval configurable :+1: